packages/dds/tree/src/shared-tree/sharedTree.ts

```diff
@@ -41,8 +41,24 @@ import {
 	IDefaultEditBuilder,
 	ForestRepairDataStore,
 } from "../feature-libraries";
+import { IEmitter, ISubscribable, createEmitter } from "../events";
 import { TransactionResult } from "../util";
 
+/**
+ * Events for {@link ISharedTreeCheckout}.
+ * @alpha
+ */
+export interface CheckoutEvents {
+	/**
+	 * A batch of changes has finished processing and the checkout is in a consistent state.
```

Comment:
The word safe is mentioned here, but it's not clearly documented what is *unsafe*. That should be clearly documented in a discoverable place. Perhaps it would also be useful to consider buffering events at the anchor set or editable tree layer to avoid unsafe reads?

Reply:
If we don't do that, should we have a "beforeBatch" event to allow consumers to disconnect events? I don't like that model, but if we leave it I'm not sure what else to do.

Reply:
See AnchorSetRootEvents and AnchorSetRootEvents for documentation of what is unsafe. Note `@remarks` mentions this.

Feel free to submit proposals to the people working on the binder API for how you want it to buffer things.

Reply:
> If we don't do that, should we have a "beforeBatch" event to allow consumers to disconnect events? I don't like that model, but if we leave it I'm not sure what else to do.

if you disconnected all your anchor events before triggering anchor events, they would be useless. As documented on those events, it is never ok to access anchors or forest from them.